File: 11-10-2025_Maximum_Total_Damage_With_Spell_Casting.java
class Solution {
    public long maximumTotalDamage(int[] power) {
         Map<Long, Long> total = new HashMap<>();
        for (int p : power) {
            total.put((long)p, total.getOrDefault((long)p, 0L) + p);
        }

        // Step 2: Sort unique damage values
        List<Long> vals = new ArrayList<>(total.keySet());
        Collections.sort(vals);
        int n = vals.size();

        // Step 3: DP array
        long[] dp = new long[n];
        dp[0] = total.get(vals.get(0));

        for (int i = 1; i < n; i++) {
            long take = total.get(vals.get(i));
            
            // find the rightmost j such that vals[j] <= vals[i] - 3
            int j = binarySearch(vals, vals.get(i) - 3);
            if (j >= 0) take += dp[j];

            dp[i] = Math.max(dp[i - 1], take);
        }

        return dp[n - 1];
    }
     private int binarySearch(List<Long> vals, long target) {
        int l = 0, r = vals.size() - 1, ans = -1;
        while (l <= r) {
            int mid = (l + r) / 2;
            if (vals.get(mid) <= target) {
                ans = mid;
                l = mid + 1;
            } else {
                r = mid - 1;
            }
        }
        return ans;
    }
}

File: 11-10-2025_Maximum_Total_Damage_With_Spell_Casting.py
class Solution:
    def maximumTotalDamage(self, power: list[int]) -> int:
        # Step 1: total damage for each unique value
        total = Counter()
        for p in power:
            total[p] += p
        
        # Step 2: sort unique damage values
        vals = sorted(total.keys())
        n = len(vals)
        
        # Step 3: DP array
        dp = [0] * n
        dp[0] = total[vals[0]]
        
        for i in range(1, n):
            take = total[vals[i]]
            
            # find rightmost index j such that vals[j] <= vals[i] - 3
            j = bisect.bisect_right(vals, vals[i] - 3) - 1
            if j >= 0:
                take += dp[j]
            
            dp[i] = max(dp[i-1], take)
        
        return dp[-1]

        
